storage/backup/backup-cli/src/coordinator/mod.rs

```diff
@@ -0,0 +1,301 @@
+            metadata_cache_opt: opt.metadata_cache_opt,
+            epoch_ending_batch_size: opt.epoch_ending_batch_size,
+            state_snapshot_interval: opt.state_snapshot_interval,
+            transaction_batch_size: opt.transaction_batch_size,
+        }
+    }
+    pub async fn run(&self) -> Result<()> {
+        // Connect to both the local Libra node and the backup storage.
+        let backup_state =
+            metadata::cache::sync_and_load(&self.metadata_cache_opt, Arc::clone(&self.storage))
+                .await?
+                .get_storage_state();
+        let db_state = self
+            .client
+            .get_db_state()
+            .await?
+            .ok_or_else(|| anyhow!("DB not bootstrapped."))?;
+        let (tx, rx) = watch::channel(db_state);
+
+        // Schedule work streams.
+        let watch_db_state = interval(Duration::from_secs(1))
+            .then(|_| self.try_refresh_db_state(&tx))
+            .boxed_local();
+
+        let backup_epoch_endings = self
+            .backup_work_stream(
+                backup_state.latest_epoch_ending_epoch,
+                &rx,
+                Self::backup_epoch_endings,
+            )
+            .boxed_local();
+        let backup_state_snapshots = self
+            .backup_work_stream(
+                backup_state.latest_state_snapshot_version,
+                &rx,
+                Self::backup_state_snapshot,
+            )
+            .boxed_local();
+        let backup_transactions = self
+            .backup_work_stream(
+                backup_state.latest_transaction_version,
+                &rx,
+                Self::backup_transactions,
+            )
+            .boxed_local();
+
+        let mut all_work = stream::select_all(vec![
+            watch_db_state,
+            backup_epoch_endings,
+            backup_state_snapshots,
+            backup_transactions,
+        ]);
+
+        loop {
+            all_work
+                .next()
+                .await
+                .ok_or_else(|| anyhow!("Must be a bug: we never returned None."))?
+        }
+    }
+}
+
+impl BackupCoordinator {
+    async fn try_refresh_db_state(&self, db_state_broadcast: &watch::Sender<DbState>) {
+        match self.client.get_db_state().await {
+            Ok(s) => db_state_broadcast
+                .broadcast(s.expect("Db should have been bootstrapped."))
+                .map_err(|e| anyhow!("Receivers should not be cancelled: {}", e))
+                .unwrap(),
+            Err(e) => println!("Failed pulling DbState from local Libra node: {}", e),
+        };
+    }
+
+    async fn backup_epoch_endings(
+        &self,
+        last_epoch_ending_epoch_in_backup: Option<u64>,
+        db_state: DbState,
+    ) -> Result<Option<u64>> {
+        let (first, last) = get_batch_range(
+            last_epoch_ending_epoch_in_backup,
+            self.epoch_ending_batch_size,
+        );
+
+        // <= because `db_state.epoch` hasn't ended yet
```

Comment:
```suggestion
        // `db_state.epoch` may not have ended yet
```

Reply:
I think we always return the current OPEN epoch, so it's not a "may"